src/main/pages/developer-guide/databases.adoc

```diff
@@ -0,0 +1,172 @@
+* PostgreSQL (version 9.6)
+
+H2 database is used by default in Single-User Che, whereas postgresql is used for Multi-user installation.
+Both of these have been tested with Selenium Integration tests, as well as TCK integration tests.
+
+It is possible to connect own DB to datasource, and have TCK tests run on it, to ensure that DB can be used with Che's JPA code.
+
+[id="datasource-components"]
+== Datasource components
+*DataSource factory*
+To Configuring Che to use custom DataSource, a new implementation of `JNDIDataSourceFactory` must be defined.
+It will configure JDBC Pool properties such as
+
+* DB username
+* DB password
+* DB URL
+* Driver class name
+* Max Total
+* Max Idle
+* Max Wait milliseconds
+
+(More about JDBC configuration properties at https://tomcat.apache.org/tomcat-8.5-doc/jndi-datasource-examples-howto.html)
+
+*ExceptionHandler*
+Database may sometimes produce different SQL exception, depending on the vendor.
+As in Java code, and TCK it is expected, a custom ExceptionHandler implementaion may be created, that will allow to rethrow more common exception.
+For example, in H2 and PostgreSQL, an exception may occur, when data couldn't be updated/stored due to unique constrain violation.
+But as it will a vendor specific exception with different code, there are Custom `H2ExceptionHandler`/`PostgreSqlExceptionHandler` that will catch it, and throw instead a `DuplicateKeyException`, which is used across the codebase.
+
+Exception handler has to be defined in persistence XML.
+
+[id="test-compatibility-kit"]
+== Test Compatibility Kit (TCK)
+Test Compatibility kit is a set of integration tests that covers uses of essential DAO components, that are being run against real database.
+
+List of Che modules that have TCK tests:
+
+* che-core-api-user
+* che-core-api-account
+* che-core-api-installer
+* che-core-api-ssh
+* che-core-api-workspace
+* che-core-api-workspace-activity
+* infrastructure-kubernetes
+* che-multiuser-machine-authentication (Multi User Che)
+* che-multiuser-api-organization (Multi User Che)
+* che-multiuser-api-permission (Multi User Che)
+* che-multiuser-api-resource (Multi User Che)
+* che-multiuser-permission-workspace (Multi User Che)
+
+Tests in these modules are executed on H2 Database during the build.
+Also, tests themselves are packaged into separate artifact with `tests` classifier, which are meant to be used for their re-usage against other DBs.
+
+As an example of that, there are `postgresql-tck` and `che-multiuser-postgresql-tck` modules for running TCK tests on Postgres DB.
+
+[id="cascade-removal-tests"]
+== JPA Cascade Removal Tests
+
+Objects in Database, upon removal, sometimes may need to be removed with other related objects.
+E.g. when Account is removed from DB, all associated Workspaces must be removed as well.
+This is achieved by Account DAO throwing a custom type of `CascadeEvent`, and Workspace DAO have a `CascadeEventSubscriber` that would listen to that event.
```

Comment:
In Eclipse Che database schema, some objects have foreign keys reference on other objects.
Example: Account and workspaces, Permissions and Workspaces. To be able to correctly
remove such objects and avoid referential integrity problems Eclipse Che are using events mechanism.
For example, when Account DAO wants to remove account it sends `CascadeEvent`, and Workspace DAO has a `CascadeEventSubscriber` that would listen to that event
and remove workspace at that time. In case if removal of workspace failed the whole transaction would be rolled back.

Reply:
ok